### PR TITLE
Request details

### LINKED ISSUE #242: Feedback
@PaulLewis was just trying out web-push and one thing that would have helped him would be to have a way to get what would be sent to the push service without actually making the request.

We currently only expose the encrypt method.

Would be good to expose the getVAPIDHeaders() and also expose a method generateRequestDetails() that returns the headers, endpoint and body that would other be used to make the final network request.

@marco-c what do you think of this?

### PR DESCRIPTION
**Pull request by gauntface**
Fixes #242 

cc @marco-c

### CODE REVIEW

**Comment by gauntface** [comment]
@marco-c  any chance of getting a minor release if this lands?

**Comment by marco-c** [comment]
> @marco-c any chance of getting a minor release if this lands?

Sure, I will look at this today and publish a minor release after I merge it.

File: README.md [hunk 1]
```diff
@@ -258,6 +258,131 @@ encryption.
...
+
+const vapidHeaders = vapidHelper.getVapidHeaders(
+  audience,
+  'mailto: example@web-push-node.org',
+  vapidDetails.publicKey,
+  vapidDetails.privateKey
+);
+```
+
+Encrypts the payload according to the [Message Encryption for Web
```
Comment (marco-c) [inline README.md]: Isn't this the description of the `encrypt` method?

File: README.md [hunk 2]
```diff
@@ -258,6 +258,131 @@ encryption.
...
+- **gcmAPIKey** can be a GCM API key to be used for this request and this
+request only. This overrides any API key set via `setGCMAPIKey()`.
+- **vapidDetails** should be an object with *subject*, *publicKey* and
+*privateKey* values defined. These values should follow the [VAPID Spec](https://tools.ietf.org/html/draft-thomson-webpush-vapid).
+- **TTL** is a value in seconds that describes how long a push message is
+retained by the push service (by default, four weeks);
+
+### Returns
+
+An object contains all the details needed to make a network request, the
```
Comment (marco-c) [inline README.md]: Nit: `An object containing`

File: README.md [hunk 3]
```diff
@@ -258,6 +258,131 @@ encryption.
...
+    pushSubscription,
+    payload,
+    options
+  );
+} catch (err) {
+  console.error(err);
+}
+```
+
+> **Note:** `generateRequestDetails()` you don't need to define a payload,
```
Comment (marco-c) [inline README.md]: Nit: could you make this sentence a little more clear?

File: src/web-push-lib.js [hunk 4]
```diff
@@ -152,29 +153,25 @@ WebPushLib.prototype.sendNotification =
...
-      } catch (err) {
-        return Promise.reject(err);
-      }
+      const encrypted = encryptionHelper.encrypt(
+        subscription.keys.p256dh, subscription.keys.auth, payload);
+
+      requestDetails.headers['Content-Length'] = encrypted.cipherText.length;
+      requestDetails.headers['Content-Type'] = 'application/octet-stream';
+      requestDetails.headers['Content-Encoding'] = 'aesgcm';
+      requestDetails.headers.Encryption = 'keyid=p256dh;salt=' +
```
Comment (marco-c) [inline src/web-push-lib.js]: Need to merge from master, `keyid` should not be here anymore.

File: README.md [hunk 5]
```diff
@@ -275,11 +275,8 @@ const vapidHeaders = vapidHelper.getVapidHeaders(
 );
 ```
 
-Encrypts the payload according to the [Message Encryption for Web
-Push](https://webpush-wg.github.io/webpush-encryption/) standard.
-
-> (*sendNotification* will automatically encrypt the payload for you, so if
-> you use *sendNotification* you don't need to worry about it).
+The *getVapidHeaders()* method will take in the values needed to create a
```
Comment (marco-c) [inline README.md]: Nit: 'a' and 'an' 😄

File: README.md [hunk 6]
```diff
@@ -335,8 +332,11 @@ try {
 }
 ```
 
-> **Note:** `generateRequestDetails()` you don't need to define a payload,
-and this method will return details without a GCM API Key and / or VAPID keys.
+> **Note:** When calling `generateRequestDetails()` the payload argument
+does not *need* to be defined, passing in null will return no body and
+> exclude any unnesscary headers.
```
Comment (marco-c) [inline README.md]: Nit: `unnesscary`

File: src/web-push-lib.js [hunk 7]
```diff
@@ -152,27 +153,25 @@ WebPushLib.prototype.sendNotification =
...
 
-        requestOptions.headers['Content-Length'] = encrypted.cipherText.length;
-        requestOptions.headers['Content-Type'] = 'application/octet-stream';
-        requestOptions.headers['Content-Encoding'] = 'aesgcm';
-        requestOptions.headers.Encryption = 'salt=' + encrypted.salt;
-        requestOptions.headers['Crypto-Key'] = 'dh=' + urlBase64.encode(encrypted.localPublicKey);
+      requestDetails.headers['Content-Length'] = encrypted.cipherText.length;
+      requestDetails.headers['Content-Type'] = 'application/octet-stream';
+      requestDetails.headers['Content-Encoding'] = 'aesgcm';
+      requestDetails.headers.Encryption = 'keyid=p256dh;salt=' +
```
Comment (marco-c) [inline src/web-push-lib.js]: `keyid` should be removed

File: src/web-push-lib.js [hunk 8]
```diff
@@ -152,27 +153,25 @@ WebPushLib.prototype.sendNotification =
...
+        encrypted.salt;
+      requestDetails.headers['Crypto-Key'] = 'keyid=p256dh;dh=' +
```
Comment (marco-c) [inline src/web-push-lib.js]: Also here

**Comment by marco-c** [comment]
LGTM, will wait for Travis before merging.

**Comment by marco-c** [comment]
I will release a new minor version shortly.

**Comment by marco-c** [comment]
3.1.0 published!

**Comment by gauntface** [comment]
Thanks marco

**Pull request merged**